### ISSUE TITLE
Improve performance of suggestion builder

### ISSUE THREAD

**Issue by TheColorRed** [HvyIndustries/crane#261]
Suggestions take over 5 seconds to appear - I assume that this is because of the size of the repo it is slow. But it takes about 5 seconds for a class to be suggested (same goes for methods/properties).

![slow auto complete](https://cloud.githubusercontent.com/assets/1571806/23827344/4a6e16a2-0677-11e7-9d9d-a9f44ccfb68c.gif)

One idea that I have that might fix this, is to have your main tree of items, then each file has its own tree specific to that file.

So, for example, I am using all of these namespaces in one file:

```php
use Laravel\Socialite\Contracts\User as ProviderUser;
use App\SocialAccount;
use App\User;
use Socialite;
```

So, now you can search through 4 classes instead of 10,000 classes. I think you could suggest faster if you just loaded those into a tree for that file.

Another idea would be to have multiple arrays that store information. For example an array that only has a list of classes, one with only traits, one with only methods, one with only properties, etc.

**Comment by nevadascout** [comment]
This is a duplicate of #220 but as this issue has more info (+ some good ideas on how to fix it) I'll close the other issue and keep this one instead.

I'm reluctant to limit the suggestions to only the included classes as that would significantly reduce the usefulness of the code completion (for me at least) as I use it when I can't remember what namespace a class is in.

I like the idea of caching information in arrays though, I started to do that before but maybe I'll look into it in much more detail in the near future.

**Comment by TheColorRed** [comment]
I think that the fastest way would not to be shrink the number of items in an array, but the memory size of the array, so splitting them into arrays with less data I think would improve performance.

I have also done some performance tests on some of my own js projects, and I found that the fastest way to loop through an array is to use a `while` loop like so (note you are looping backwards this way it also allows for deleting of items in the loop if needed):

```js
let i = items.length;
while(i--) {
    let item = items[i];
    // Use the item
}
```

The second fastest way is to use a for with caching.

```js
for(let i = 0, l = items.length; i < l; i++){} // Fast

// This is slower because it doesn't cache the number of items
for(let i = 0; i < items.length; i++){} // Slow
```

This however, didn't seem to be as fast, as I seemed to gain noticeable speeds with the while loop.

**Comment by TheColorRed** [comment]
Objects vs strings in loops: http://jsben.ch/#/8kub4
Speed of loops: http://jsben.ch/#/GnGON

**Comment by nevadascout** [comment]
I've split out #262 to replace all of the `forEach` loops in the suggestion builder with cached `for` loops instead. This can be done for the v0.3.3 release, then full optimisation (with new arrays, etc) will be done for the v0.3.4 or v0.3.5 releases later this week.

**Comment by ichiriac** [comment]
I've put some console output into crane action in order to track what is called.

At the end of a little file I've just typed `$`, here are the results in magento2 ; 

```
[INFO] Sun Mar 19 2017 11:19:12 GMT+0100 (CET) : Request completion
[INFO] File Changed: /home/devbox/Documents/glayzzle/php-parser-dev-tokens/framework/magento2/app/code/Magento/AdminNotification/Controller/Adminhtml/Notification.php
[INFO] Sun Mar 19 2017 11:19:12 GMT+0100 (CET) : Get Scope
[INFO] Sun Mar 19 2017 11:19:12 GMT+0100 (CET) : Start build completion : $
[INFO] Sun Mar 19 2017 11:19:12 GMT+0100 (CET) : Found 17219 items / remove duplicates
[INFO] Sun Mar 19 2017 11:19:16 GMT+0100 (CET) : Resulting 9148 items
[INFO] Sun Mar 19 2017 11:19:16 GMT+0100 (CET) : Request parsing : /home/devbox/Documents/glayzzle/php-parser-dev-tokens/framework/magento2/app/code/Magento/AdminNotification/Controller/Adminhtml/Notification.php
[INFO] Sun Mar 19 2017 11:19:16 GMT+0100 (CET) : Parsing done, indexing
[INFO] Sun Mar 19 2017 11:19:16 GMT+0100 (CET) : Request parsing : /home/devbox/Documents/glayzzle/php-parser-dev-tokens/framework/magento2/app/code/Magento/AdminNotification/Controller/Adminhtml/Notification.php
[INFO] Sun Mar 19 2017 11:19:16 GMT+0100 (CET) : Parsing done, indexing
[INFO] Sun Mar 19 2017 11:19:16 GMT+0100 (CET) : Request parsing : /home/devbox/Documents/glayzzle/php-parser-dev-tokens/framework/magento2/app/code/Magento/AdminNotification/Controller/Adminhtml/Notification.php
[INFO] Sun Mar 19 2017 11:19:16 GMT+0100 (CET) : Parsing done, indexing
[INFO] Sun Mar 19 2017 11:19:16 GMT+0100 (CET) : Index done
[INFO] Sun Mar 19 2017 11:19:16 GMT+0100 (CET) : Index done
[INFO] Sun Mar 19 2017 11:19:16 GMT+0100 (CET) : Index done
[INFO] Processing complete!
```

Here with new keyword (without noise about parsing again the file) : 

```
[INFO] Sun Mar 19 2017 11:25:12 GMT+0100 (CET) : Parsing done, indexing
[INFO] Sun Mar 19 2017 11:25:12 GMT+0100 (CET) : Request completion
[INFO] Sun Mar 19 2017 11:25:12 GMT+0100 (CET) : Get Scope
[INFO] Sun Mar 19 2017 11:25:12 GMT+0100 (CET) : Start build completion : $foo = new A
[INFO] Sun Mar 19 2017 11:25:12 GMT+0100 (CET) : Found 15996 items / remove duplicates
[INFO] Sun Mar 19 2017 11:25:16 GMT+0100 (CET) : Resulting 8220 items
```

The `suggestionBuilder.ts` file takes > 90% of time processing on duplicates removal : 

```
for (var i = 0, l = toReturn.length; i < l; i++) {
            var item = toReturn[i];

            var found = false;

            for (var j = 0, sl = filtered.length; j < sl; j++) {
                var subItem = filtered[j];

                if (subItem.label == item.label) {
                    found = true;
                }
            }

            if (!found) {
                filtered.push(item);
            }
        }
```

To optimise this, put a break after `found = true;`, but without indexation, the iteration cost is too high (about toReturn.length * filtered.length = 130 millions inner loop)

You could make an object cache, something like this : 

```
        var filtered = [];
        let index = {};
        for (var i = 0, l = toReturn.length; i < l; i++) {
            var item = toReturn[i];
            if (!(item.label in index)) {
              filtered.push(item);
              index[item.label] = true;
            }
        }
```

Same case after optimisation (almost no time spent) :

```
[INFO] Sun Mar 19 2017 11:34:39 GMT+0100 (CET) : Request completion
[INFO] Sun Mar 19 2017 11:34:39 GMT+0100 (CET) : Get Scope
[INFO] Sun Mar 19 2017 11:34:39 GMT+0100 (CET) : Start build completion : $foo = new A
[INFO] Sun Mar 19 2017 11:34:39 GMT+0100 (CET) : Found 15996 items / remove duplicates
[INFO] Sun Mar 19 2017 11:34:39 GMT+0100 (CET) : Resulting 8220 items
```

**Comment by ichiriac** [comment]
## Improving writes on caching system :

Actually we flush all the cache into a single file. On big projects, that means every 5 sec after a change on a file, we have a flush over all the tree.

- 5 secs delay does not guarantee consistency of the tree in case of failure.
- By flushing all the tree, 5 secs may be a too short interval
- The tree may be large so I will use CPU & Disk intensively

We could fix this by :

- [ ] Separating ast symbols, one file file
- [ ] As flushing little files, maybe disabling zip support (no more need to use CPU for zip/unzip)

## Avoid parsing too early

Same trick as the cache flushing system, we could wait and trigger parsing every X seconds, instead at each document change, resulting in a decrease of CPU usage. Will be a good practice later for document validation.

**Comment by nevadascout** [comment]
I've implemented a basic caching system in the suggestion builder for the 0.3.4 release, and I have created a new task (#271) to address the performance issues with the file caching and parsing.

**Comment by nevadascout** [comment]
Released in v0.3.4
